vision/segmentation/er_ccl_clustering.go

```diff
@@ -0,0 +1,289 @@
+func (erCCL *ErCCLConfig) ConvertAttributes(am utils.AttributeMap) error {
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{TagName: "json", Result: erCCL})
+	if err != nil {
+		return err
+	}
+	err = decoder.Decode(am)
+	if err == nil {
+		err = erCCL.CheckValid()
+	}
+	return err
+}
+
+// NewERCCLClustering returns a Segmenter that removes the ground plane and returns a segmentation
+// of the objects in a point cloud using a connected components clustering algo described in the paper
+// "A Fast Spatial Clustering Method for Sparse LiDAR Point Clouds Using GPU Programming" by Tian et al. 2020.
+func NewERCCLClustering(params utils.AttributeMap) (Segmenter, error) {
+	// convert attributes to appropriate struct
+	if params == nil {
+		return nil, errors.New("config for ER-CCL segmentation cannot be nil")
+	}
+	cfg := &ErCCLConfig{}
+	err := cfg.ConvertAttributes(params)
+	if err != nil {
+		return nil, err
+	}
+	return cfg.ErCCLAlgorithm, nil
+}
+
+// ErCCLAlgorithm applies the connected components clustering algorithm directly on a given point cloud.
+func (erCCL *ErCCLConfig) ErCCLAlgorithm(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+	// get next point cloud
+	cloud, err := src.NextPointCloud(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// run ransac, get pointcloud without ground plane
+	ps := NewPointCloudGroundPlaneSegmentation(cloud, erCCL.MaxDistFromPlane, erCCL.MinPtsInPlane, erCCL.AngleTolerance, erCCL.NormalVec)
+	// if there are found planes, remove them, and keep all the non-plane points
+	_, nonPlane, err := ps.FindGroundPlane(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// need to figure out coordinate system
+	// if height is not y, then height is going to be z
+	heightIsY := erCCL.NormalVec.Y != 0
+
+	// calculating s value, want 200 x 200 graph
+	resolution := math.Ceil((nonPlane.MetaData().MaxX - nonPlane.MetaData().MinX) / 200)
+	if heightIsY {
+		resolution = math.Ceil((math.Ceil((nonPlane.MetaData().MaxZ-nonPlane.MetaData().MinZ)/200) + resolution) / 2)
+	} else {
+		resolution = math.Ceil((math.Ceil((nonPlane.MetaData().MaxY-nonPlane.MetaData().MinY)/200) + resolution) / 2)
+	}
+
+	// create obstacle flag map, return that 2d slice of nodes
+	labelMap := pcProjection(nonPlane, resolution, heightIsY)
+
+	// actually run erCCLL
+	// iterate through every box, searching down and right r distance
+	// run calculations to meet similarity threshold
+	// if similar enough update to initial label value (will also be smallest)
+	// iterate through pointcloud
+
+	i := 0
+	continueRunning := true
+	for continueRunning {
+		// 0.9 is alpha
+		continueRunning := labelMapUpdate(labelMap, erCCL.ClusteringRadius, 0.9, erCCL.ClusteringStrictness, resolution)
+		if !continueRunning {
+			break
+		}
+
+		if i > MaxCCLIterations { // arbitrary cutoff for iterations
+			return nil, errors.New("could not converge, change parameters")
+		}
+		i++
+	}
+
```

Comment:
Speaking of abstraction, could you take all the code having to do with creating/updating the labelMap, and put it in its own public function? We might want to apply the CCL clustering steps on the obstacles_depth algorithm, too. So imagine you have the initial labelMap already, and you feed that into the function, and get back the final label map.